pkg/background/update_request_controller.go

```diff
@@ -298,11 +258,66 @@ func (c *Controller) deleteUR(obj interface{}) {
 			return
 		}
 	}
+	// sync Handler will remove it from the queue
+	c.enqueueUpdateRequest(ur)
+}
 
-	if ur.Status.Handler != "" {
-		return
+func (c *Controller) processUR(ur *kyvernov1beta1.UpdateRequest) error {
+	switch ur.Spec.Type {
+	case kyvernov1beta1.Mutate:
+		ctrl, _ := mutate.NewMutateExistingController(c.kyvernoClient, c.client,
+			c.policyLister, c.npolicyLister, c.urLister, c.eventGen, logger, c.configuration)
+		return ctrl.ProcessUR(ur)
+
+	case kyvernov1beta1.Generate:
+		ctrl, _ := generate.NewGenerateController(c.kyvernoClient, c.client,
+			c.policyLister, c.npolicyLister, c.urLister, c.eventGen, c.nsLister, logger, c.configuration,
+		)
+		return ctrl.ProcessUR(ur)
 	}
+	return nil
+}
 
-	// sync Handler will remove it from the queue
-	c.enqueueUpdateRequest(ur)
+func (c *Controller) markUR(ur *kyvernov1beta1.UpdateRequest) (*kyvernov1beta1.UpdateRequest, bool, error) {
+	ur = ur.DeepCopy()
+	var ok bool
+	err := retry.RetryOnConflict(common.DefaultRetry, func() error {
+		var retryError error
+		if ur.Status.Handler != "" {
```

Comment:
@realshuting we need to move this check inside the retry i think.

Reply:
Good catch! Instead of moving it, can we add this as the additional check? So that we can return earlier without updating `.status`.

Reply:
Arf, we also need to fetch the UR again ourselves too :(

Reply:
Where do we need to fetch it again exactly? In `RetryOnConflict`?

The goal is to mark the UR with the Kyverno pod's name and let the current instance reconcile it. This is to avoid duplicate processing in HA. What would be the best way to handle such a scenario?

Reply:
Looked at the docs https://pkg.go.dev/k8s.io/client-go/util/retry#RetryOnConflict
The call to `UpdateStatus` does not return the updated resource.

Reply:
Isn't the updated UR returned from this call? https://github.com/kyverno/kyverno/pull/3964/files#diff-2a4d1510c36747a184576e3854dea756e5871ec942296b0e2c00022f99fe97beR292

Reply:
> Where do we need to fetch it again exactly? In RetryOnConflict?

Yes

> The goal is to mark the UR with the Kyverno pod's name and let the current instance reconcile it. This is to avoid duplicate processing in HA. What would be the best way to handle such a scenario?

I understand why we are doing this, I'm not sure what the best solution is in this case. Maybe leader election ?

Reply:
> Isn't the updated UR returned from this call? https://github.com/kyverno/kyverno/pull/3964/files#diff-2a4d1510c36747a184576e3854dea756e5871ec942296b0e2c00022f99fe97beR292

It looks like it isn't. From the docs:
```
    // Fetch the resource here; you need to refetch it on every try, since
    // if you got a conflict on the last update attempt then you need to get
    // the current version before making your own changes.
```

Reply:
We disabled the leader election specifically for this controller as we want to distribute works across all instances. This also reduces the `Get` Client calls per instance to minimize the chances we get throttled.

Reply:
Another option would be to use some kind of sharding but it would be a significant change.

Reply:
> Looked at the docs https://pkg.go.dev/k8s.io/client-go/util/retry#RetryOnConflict The call to `UpdateStatus` does not return the updated resource.

Makes sense! Then let's get it again using ur lister.

Reply:
Yes, that's what i did :crossed_fingers:

Reply:
Now it fails in the policy controller.